### ISSUE TITLE
Make LB configuration optional

### ISSUE THREAD

**Issue by johananl** [kinvolk/lokomotive#372]
I think it should be possible to decide not to enable ingress for certain worker pools. For example, it is likely that users will deploy worker pools which run only backing services (e.g. DBs and queues).

I suggest we add a config knob at the worker pool level which determines whether LB resources are provisioned for that pool. This is platform-specific: on AWS, for example, this knob would control whether ELB listeners and target groups are created for the pool. On Packet, this know would determine whether BGP gets enabled for the nodes in the pool.

We should implement this feature in a way which exposes a uniform knob (or knobs) to the user, regardless the platform. The UX should be the same regardless the implementation details. It is likely that *some* of the mechanisms for achieving this functionality are already in place, however we should still verify things are uniform and make sense at the UX level.

Lastly, there is the components layer: it probably doesn't make sense to deploy Contour and cert-manager when the LB knob is "off", and it doesn't make sense *not* to deploy Contour when the knob is "on". For now it can suffice to mention in the docs that changing this know should be followed by a deployment/removal of the `contour` component, however we may want to think about coupling these two separate operations into one.

**Comment by rata** [comment]
> I think it should be possible to decide not to enable ingress for certain worker pools. For example, it is likely that users will deploy worker pools which run only backing services (e.g. DBs and queues).

Can you please elaborate on why you think it should be possible. And what is the use case ti will enable or improve, exactly?

I assume we are all aware, but just in case: the general rule for most kubernetes deployments is that  all your nodes are backends for a Load Balancer, kube-proxy routes incoming packets to the pods to keep their load-balanced. _This_ is what happens in a cloud environment like AWS when you create a service type LoadBalancer, in fact. (Not sure if Ingress is doing something different, haven't used it in clouds. My guess is that it is not).  And if you need source IP preservation, you use `externalTrafficPolicy: Local`.

So, while possible... is definitely not the default and not sure why we should add that functionality. Can you please elaborate? :)

> On Packet, this know would determine whether BGP gets enabled for the nodes in the pool.

This is trickier than one would think. On Packet, BGP ECMP is only till the ToR switch. So, if you have 4 servers in one rack A and 1 server in another B, the server on rack B will receive 50% of the traffic. The only reasonable way we managed to use this on packet is using hardware reservations to make sure that "ingress" nodes are evenly balanced across racks.

So... on Packet we are doing it, in some rudimentary way, but because of limitations of the platform. Is not a desired state that we wanted to achieve, but the other way around: limitations in Packet lead us to use such a setup.

**Comment by johananl** [comment]
I guess I wasn't clear about my rationale: I want to allow users to designate a group of nodes as "ingress nodes", namely nodes through which traffic enters the cluster. Assuming we support an arbitrary number of node pools (which we do), and assuming the only reason to create multiple worker pools is that the pools are different from one another *somehow* (e.g. type of hardware or security policy), it only makes sense to me that not all worker pools should automatically have load balancing + ingress enabled.

A sample use case could be a "data worker pool": a pool which hosts backing services such as databases, queues and caches and which leverages storage-optimized machines. Such a pool would need to receive traffic only from other services within the cluster, therefore it would be both unnecessary and a security risk to put a load balancer in front of this pool.

To simplify things, it makes sense to me to only enable load balancing (i.e. NLB or BGP + MetalLB, depending on the platforms) on nodes which run Contour pods.

Am I missing something?

**Comment by invidian** [comment]
I agree with @johananl, with cluster larger than few nodes, it's reasonable to have dedicated ingress nodes, which are allowed to receive traffic from the internet. I think having an option to disable LB is reasonable. 

Perhaps even LB should be disabled by default (as intention is to enable it only on one pool, instead of disabling it on every single worker pool you have), but this we can discuss separately.

**Comment by johananl** [comment]
@rata I've read your comment again and I'm not sure we're talking about the same thing. I'll do my best to clarify.

My main point is - **we shouldn't create LB resources when they are not needed**. Now, the question is "when are LB resources unnecessary?" The answer is "when my pods don't need to receive traffic from the internet".
Now, since Lokomotive supports multiple worker pools per cluster, it is likely that in some cases a dedicated worker pool would be created for backing services, namely services which shouldn't be exposed to the internet directly. Such a worker pool would serve only backend pods through the use of node selectors.
My concern is that right now, we always provision LB resources for every worker pool on AWS. On Packet, we have a `disable_bgp` knob but BGP is enabled by default. This issue is about two things at the practical level:

1. Mirror the functionality we have on Packet in AWS for the sake of feature parity.
1. Make the configuration surface identical among all platforms. The UX should feel similar.

As an example, this is what it could look like:

```hcl
cluster "aws" {
  worker_pool "data" {
    lb = false
  }
  worker_pool "web" {
    lb = true
  }
}
```

Lastly, I am well aware of how LB functions on both AWS and Packet, however I fail to see how what you wrote relates to the issue we're discussing.

>I assume we are all aware, but just in case: the general rule for most kubernetes deployments is that all your nodes are backends for a Load Balancer, kube-proxy routes incoming packets to the pods to keep their load-balanced. This is what happens in a cloud environment like AWS when you create a service type LoadBalancer, in fact. (Not sure if Ingress is doing something different, haven't used it in clouds. My guess is that it is not). And if you need source IP preservation, you use externalTrafficPolicy: Local.

All true. I don't see how the above conflicts with being able to turn load balancing on/off per worker pool. To emphasize, I'm **not** talking about turning LB off for **one node** within a pool.

>This is trickier than one would think. On Packet, BGP ECMP is only till the ToR switch. So, if you have 4 servers in one rack A and 1 server in another B, the server on rack B will receive 50% of the traffic. The only reasonable way we managed to use this on packet is using hardware reservations to make sure that "ingress" nodes are evenly balanced across racks.

Again - all true, but I don't understand how this is related to the discussion at hand. How does this support or oppose the idea of being able to not enable LB for a pool?
